Core/include/eASTMagneticField.hh

```diff
@@ -0,0 +1,227 @@
+    EInterpolationType fInterpolationType{kLinear};
+
+  public:
+
+    // Interpolation type setters and getters
+    void SetInterpolationType(const G4String& type) {
+      if (type == "linear") fInterpolationType = kLinear;
+      if (type == "cubic") fInterpolationType = kCubic;
+    }
+    EInterpolationType GetInterpolationType() const {
+      return fInterpolationType;
+    }
+
+  private:
+
+    // Linear interpolation functions
+    double _linearInterpolate(const double p[2 << 0], double x) const {
+      return p[0] + x * (p[1] - p[0]);
+    }
+    double _bilinearInterpolate(const double p[2 << 1], const double x[2]) const {
+      double c[2];
+      c[0] = _linearInterpolate(&(p[0]), x[1]);
+      c[1] = _linearInterpolate(&(p[2]), x[1]);
+      return _linearInterpolate(c, x[0]);
+    }
+    double _trilinearInterpolate(const double p[2 << 2], const double x[3]) const {
+      double c[2];
+      c[0] = _bilinearInterpolate(&(p[0]), &(x[1]));
+      c[1] = _bilinearInterpolate(&(p[4]), &(x[1]));
+      return _linearInterpolate(c, x[0]);
+    }
+
+    // Cubic interpolation functions
+    double _cubicInterpolate(const double p[4 << 0], double x) const {
+      return p[1] + 0.5 * x * (p[2] - p[0] +
+                               x * (2. * p[0] - 5. * p[1] + 4. * p[2] - p[3] +
+                                    x * (3. * (p[1] - p[2]) + p[3] - p[0])));
+    }
+    double _bicubicInterpolate(const double p[4 << 1], const double x[2]) const {
+      double c[4];
+      c[0] = _cubicInterpolate(&(p[0]),  x[1]);
+      c[1] = _cubicInterpolate(&(p[4]),  x[1]);
+      c[2] = _cubicInterpolate(&(p[8]),  x[1]);
+      c[3] = _cubicInterpolate(&(p[12]), x[1]);
+      return _cubicInterpolate(c, x[0]);
+    }
+    double _tricubicInterpolate(const double p[4 << 2], const double x[3]) const {
+      double c[4];
+      c[0] = _bicubicInterpolate(&(p[0]),  &(x[1]));
+      c[1] = _bicubicInterpolate(&(p[16]), &(x[1]));
+      c[2] = _bicubicInterpolate(&(p[32]), &(x[1]));
+      c[3] = _bicubicInterpolate(&(p[48]), &(x[1]));
+      return _cubicInterpolate(c, x[0]);
+    }
+
+};
+
+class eASTMagneticField : public G4MagneticField {
+
+  public:
+
+    eASTMagneticField();
+    ~eASTMagneticField() = default;
+
+    // Activate in ConstuctSDandField
+    void Activate();
+
+    // Create new field
+    void CreateField(const G4String& name);
+
+    // Get field value at point
+    void GetFieldValue(const G4double point[4], G4double *field) const {
+      field[0] = 0;
+      field[1] = 0;
+      field[2] = 0;
+      for (auto &map: fMaps) {
+        map.AddFieldValue(point, field);
+      }
+    };
+
+    // Print field value at point
+    void PrintFieldValue(const G4ThreeVector& point) {
+      for (auto &map: fMaps) {
+        map.PrintFieldValue(point);
+      }
+    };
+
+  private:
+
+    // Field messenger
+    G4GenericMessenger fFieldMessenger;
+
+  private:
+
+    // Field maps
+    std::vector<eASTMagneticFieldMap> fMaps;
```

Comment:
With large fieldmaps, this could be made global static. Writes to fMaps happen before threads are spun off so the mutex locks are never going to block anything, and it have all threads access the same maps in memory in thread safe const functions.